modules/gapi/src/backends/oak/goakbackend.cpp

```diff
@@ -0,0 +1,616 @@
+    {
+        // FIXME: change the hard-coded behavior (XLinkIn path)
+        auto camRgb = m_pipeline->create<dai::node::ColorCamera>();
+        // FIXME: extract camera compile arguments here and properly convert them for dai
+        camRgb->setBoardSocket(dai::CameraBoardSocket::RGB);
+        camRgb->setResolution(dai::ColorCameraProperties::SensorResolution::THE_1080_P);
+
+        // Set camera output. Fixme: consider working with other camera outputs
+        m_camera_input = camRgb;
+        // FIXME: change when other camera censors are introduced
+        m_camera_wh = camRgb->getVideoSize();
+
+        // Prepare XLinkOut nodes for each output object in graph
+        for (size_t i = 0; i < outs_data.size(); ++i) {
+            auto xout = m_pipeline->create<dai::node::XLinkOut>();
+            std::string xout_name = "xout" + std::to_string(i);
+            m_out_queue_names.push_back(xout_name);
+            xout->setStreamName(xout_name);
+            m_xlink_outputs.push_back(xout);
+        }
+
+        // Create OAK node for each node in this backend
+        for (const auto& nh : nodes) {
+            if (m_gm.metadata(nh).get<NodeType>().t == NodeType::OP) {
+                const auto& op = m_gm.metadata(nh).get<Op>();
+                const auto &u = ConstOAKGraph(m_g).metadata(nh).get<OAKComponent>();
+                // pass kernel input args and compile args to prepare OAK node and
+                // store it to link later
+                m_oak_nodes[nh] = nullptr;
+                m_oak_node_inputs[nh] = {};
+                m_oak_node_outputs[nh] = {};
+
+                std::vector<cv::GArg> in_ctx_args;
+                in_ctx_args.reserve(op.args.size());
+                for (auto &op_arg : op.args) in_ctx_args.push_back(packArg(op_arg,
+                                                                           m_oak_node_inputs[nh],
+                                                                           m_oak_wrapped_args));
+                for (auto &&op_out : op.outs) outArg(op_out, m_oak_node_outputs[nh]);
+                GAPI_Assert(!m_oak_node_inputs[nh].empty());
+                GAPI_Assert(!m_oak_node_outputs[nh].empty());
+
+                GOAKContext ctx(m_pipeline, m_camera_wh, in_ctx_args, m_oak_node_outputs[nh]);
+                u.k.m_f(ctx, m_oak_nodes[nh], m_in_queues);
+                GAPI_Assert(m_oak_nodes[nh] != nullptr);
+            }
+        }
+
+        for (const auto& el : m_oak_node_inputs) {
+            for (const auto& in : el.second) {
+                GAPI_Assert(in != nullptr);
+            }
+        }
+
+        for (const auto& el : m_oak_node_outputs) {
+            for (const auto& out : el.second) {
+                GAPI_Assert(out != nullptr);
+            }
+        }
+
+        // Properly link all nodes
+        // 1. Link input nodes to camera
+        for (const auto& d : ins_data)
+        {
+            for (const auto& nh : nodes)
```

Comment:
this loop can be optimized if you identify our input operations early (e.g. in the above loop). Just store this list elsewhere.

Reply:
Didn't get the idea

Reply:
i think the idea is remember inputs in some list at construction phase (not here but in early place) then iterate here on this list instead whole nodes and check every node type

Reply:
You can look at `d->outNodes()` only and check if it contains OAKComponent

Reply:
Any progress here?